20220714-lookup-api.md

```diff
@@ -0,0 +1,461 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+The Lookup API will provide clients with the ability to query the object ids that a user has a particular relationship with. The list of object ids the Lookup API returns can then be used by a client to filter an object collection to only the objects that a user has a particular relationship with by taking the intersection of it with any client provided filters/ordering of the underlying dataset. This is described in more detail with [Search with Permission (Option 3)](https://openfga.dev/docs/interacting/search-with-permissions#option-3-build-a-list-of-ids-then-search) document.
+
+It will be provided as two different API endpoints: one will provide a streaming endpoint and the other will implement the endpoint without streaming. These two different endpoints could be used interchangably, but clients can choose which one may be better for their integration environment.
+
+Consider a developer/client integrating with OpenFGA with the following authorization model:
+```
+type folder
+    relations
+        define viewer as self
+
+type document
+    relations
+        define viewer as self or editor or viewer from parent
+        define editor as self
+        define parent as self
+```
+and the following relationship tuples:
+
+| object         | relation | user           |
+|----------------|----------|----------------|
+| document:doc1  | viewer   | bob            |
+| document:doc2  | editor   | bob            |
+| document:doc3  | parent   | folder:folder1 |
+| folder:folder1 | viewer   | bob            |
+
+## StreamedLookup
+StreamedLookup will implement a grpc [server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) endpoint.
+
+The endpoint will stream each object id the target user has the given relationship with as it is evaluated, and the server will continuing streaming results until either the lookup has been exhausted or the `streamed-lookup-deadline` has been reached (see [Configuration Changes](#configuration-changes)).
+
+Sample grpcurl command:
+```
+grpcurl -plaintext \
+  -d '{
+           "store_id":"<store>", 
+           "authorization_model_id":"<modelID>",
+           "object_type":"document", 
+           "relation":"viewer", 
+           "user":"bob"
+      }' \
+  localhost:8081 openfga.v1.OpenFGAService/StreamedLookup
+
+{
+    "object_id": "doc1",
+}
+{
+    "object_id": "doc2"
+}
+{
+    "object_id": "doc3"
+}
+```
+
+Sample curl command:
+```
+curl --request POST 'http://localhost:8080/stores/<storeID>/lookup:streamed' \
```

Comment:
Having ":" in the path seems a bit inconsistent with the rest of the APIs.  Will it make more sense if streamed is pass in as an optional body parameter?

Reply:
We don't have support for proxying an RPC based on a optional body (or query) parameter. We use grpc-gateway and it is limited to a direct HTTP path mapping to the underlying RPC.

So our best option at this time is to encode the different RPC endpoints in the HTTP path. We could do something different like `localhost:8080/stores/<storeID>/streamedLookup` 🤷

Reply:
I'd stick to something like `streamedLookup`

Reply:
Since we've chosen to rename this API endpoint to `ListObjects`, then for the streaming variant I will use the name `/streamedListObjects` to go along with your recommendations. I've made this change.

Reply:
In that case, maybe `streamed-lookup`?

Consistent with `read-changes`. And we don't generally use capitalization in path names